typescript.md

```diff
@@ -0,0 +1,283 @@
+
+
+## Use type assertions to `unknown` when the going gets tough
+
+At times, you may hit issues because of type mismatching.
+
+If you are certain that the implementation is correct (but maybe the typing is only partially correct),
+rather than fall back to `// @ts-ignore`, use a type assertion to unknown and then to the intended type.
+
+```typescript
+type Person = {
+  name: string;
+  gender: string;
+};
+
+let fn = function(): Person {
+  let p = {} as Partial<Person>;
+  p.name = 'Luiza';
+  p.gender = 'unspecified';
+
+  // will complain that Partial<Person> does not match Person
+  // return p;
+
+  // since you are certain that p will actually have all the required properties, you can
+  // return p as unknown as Person;
+};
+```
+
+## Don't forget mapped types generics
+
+Generics are a good for mapping types
+e.g. `type MaybePromise<T> = T | Promise<T>` allows for `let a: MaybePromise<void>`, `let b: MaybePromise<number>`, etc.
+
+But they are also good for creating type inferences in function signatures e.g.
+
+```typescript
+let identity = function<T>(a: T): T {
+  return a;
+};
+
+// PS: Naïve example. TypeScript would have perfectly inferred the type, yes.
+```
+
+
+## Use `unknown` as much as possible, and not `any`
+
+`unknown` is the perfect default type, when you cannot proxy types
+
+Whenever an `unknown` variable will be used (e.g. access a property) or passed onto a function with a type constraint,
+typescript will force you to cast it, and thus keep your code safely typed. `any` on the otherhand would allow anything.
+
+[Read more about the unknown type.](https://mariusschulz.com/blog/the-unknown-type-in-typescript)
+
+
+## Use utility types
+
+Newer versions of TypeScript have opened up the possibility for so-called utility types e.g.
+
+* `Partial<T>` would be a type for a partial object of type T
+* `ReturnType<T>` would be the type of the return value of a function of type T
+
+Similarly to the built-in types, more utility-types exist as external packages:
+  * our own collection: https://github.com/tobiipro/lodash-firecloud/blob/master/src/types.ts
+  * built-in utility types: https://codewithstyle.info/Comprehensive-list-of-useful-built-in-types-in-TypeScript/
+  * extension of utility types: https://github.com/piotrwitek/utility-types
+  * **comprehensive** but can easily through TypeScript errors because of high complexity: https://github.com/pirix-gh/ts-toolbelt
+
+[Read more about utility types here.](https://www.typescriptlang.org/docs/handbook/utility-types.html).
+
+
+## Amend modules and interfaces (and classes)
+
+Sometimes external modules have incorrect types, and you need to amend their interfaces
+
+or to amend classes e.g.
+
+```typescript
+interface ClockInterface {
+    currentTime: Date;
+}
+
+class Clock implements ClockInterface {
+  currentTime: Date = new Date();
+  constructor(h: number, m: number) { }
```

Comment:
Sorry. I don't get what this code should show and thus not why we amend the prototype. The constructor does nothing?? 

Perhaps the hour is too late...